src/runtime/virtcontainers/container.go

```diff
@@ -672,6 +677,28 @@ func (c *Container) createBlockDevices(ctx context.Context) error {
 			continue
 		}
 
+		// Check for scenario where CSI is providing a FS based volume which is not mounted on the host. In this
+		// scenario,a special file will be persent within the mount's source which provides details on where the
+		// backing block device comes from. In this scenario, we ignore the mount on the host, and instead pass
+		// the block device to the VM, and mount it in the guest to provide to the container workload.
+		if _, err := os.Stat(filepath.Join(m.Source, directvolume.DirectAssignedVolumeJson)); err == nil {
+			c.Logger().Debugf("found direct assigned volume: %s", m.Source)
+
+			directInfo, err := getDirectAssignedDiskMountInfo(filepath.Join(m.Source, directvolume.DirectAssignedVolumeJson))
+			if err != nil {
+				c.Logger().WithError(err).Error("direct assigned volume found, but error reading file")
+				return err
+			}
+
+			// Update the volume information based on what we received from CSI:
+			c.mounts[i].Source = directInfo.Device
+			c.mounts[i].Type = directInfo.FsType
+			m.Source = directInfo.Device
+			m.Type = directInfo.FsType
+
+			c.mounts[i].Options = strings.Split(directInfo.Options, ",")
+			m.Options = c.mounts[i].Options
+		}
```

Comment:
should here add a `continue` to skip non-direct volumes?

Reply:
No, we want to fall through and handle. In our case, it'll fall through at then identify that the mount is a block device file, and create the appropriate DeviceInfo struct, and pass it on to the device manager.